### PR TITLE
Let user decide nodes platforms and roles. (+ fix for aliases)

### PR DESCRIPTION
**Pull request by jsuchome**
Added options to set intended roles and target OS (platform) of the nodes.

Tested with

    want_node_roles=controller=2,compute=2,storage=3
    want_node_os=suse-12.1=4,suse-12.0=3
    want_node_aliases=controller=2,compute=2,ceph=3

### CODE REVIEW

File: scripts/qa_crowbarsetup.sh [hunk 1]
```diff
@@ -1648,35 +1648,55 @@ function onadmin_allocate()
...
-        local nodes=(
-            $(get_all_discovered_nodes | head -n 3)
-        )
-        for n in $(seq 1 2); do
-            echo "Setting node $(($n+1)) to SLE12 Storage..."
-            set_node_role_and_platform ${nodes[$n]} "storage" "suse-12.0"
+            i=1
+            for node in `printf  "%s\n" $nodesavailable | head -n$number`; do
+                # FIXME or keep this to platform only?
+                local $role = "compute"
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: Hrm, so how do we get controllers? :-)

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: Exactly. Maybe this function should only set the platform, and we'd need another one with the same syntax, assigning roles. Like "want_node_roles=controller=2,storage=3,compute=1".

Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: Maybe, yes.

File: scripts/qa_crowbarsetup.sh [hunk 2]
```diff
@@ -1562,6 +1572,32 @@ function set_node_role_and_platform()
...
+
+# set a node's platform
+# must be run after discovery and before allocation
+function set_node_platform()
+{
+    node="$1"
+    platform="$2"
+    local t=$(mktemp).json
+    knife node show -F json "$node" > $t
+    json-edit $t -a normal.target_platform -v "$platform"
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: I dislike the code-duplication. IMHO there should be one function that gets a 3rd param that is either normal.target_platform or normal.crowbar_wall.intended_role

**Comment by jdsn** [comment]
this will be obsolete with the usage of crowbar batch, right?

So I'd rather invest into the usage of crowbar batch rather than building up more mkcloud workarounds that do the same thing (and will be dropped anyway soon)

File: scripts/qa_crowbarsetup.sh (hunk 2 above)
Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: I understand that feeling, but I think the readability is better current way...

**Comment by jsuchome** [comment]
@jdsn No, crowbar batch cannot set roles and/or target platform

**Comment by jsuchome** [comment]
Tested again, with manual mkcloud run and the options above.

**Comment by bmwiedemann** [comment]
:+1:

File: scripts/qa_crowbarsetup.sh [hunk 3]
```diff
@@ -1547,21 +1557,37 @@ function onadmin_installcrowbar()
...
     node="$1"
-    role="$2"
-    platform="$3"
+    attr="$2"
+    value="$3"
     local t=$(mktemp).json
     knife node show -F json "$node" > $t
-    json-edit $t -a normal.crowbar_wall.intended_role -v "$role"
-    json-edit $t -a normal.target_platform -v "$platform"
+    json-edit $t -a "normal.$2" -v "$value"
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: s/$2/$attr/

File: scripts/qa_crowbarsetup.sh [hunk 4]
```diff
@@ -1547,21 +1557,37 @@ function onadmin_installcrowbar()
...
-function set_node_role_and_platform()
+# Set a node's attribute (see 2nd argument)
+# Must be run after discovery and makes sense mostly before allocation
+function set_node_attribute()
 {
     node="$1"
-    role="$2"
-    platform="$3"
+    attr="$2"
+    value="$3"
```
Comment (aplanas) [inline scripts/qa_crowbarsetup.sh]: Are those variables `local`?

File: scripts/qa_crowbarsetup.sh [hunk 5]
```diff
@@ -1648,35 +1674,67 @@ function onadmin_allocate()
...
-        )
+        local nodesavailable=`get_all_discovered_nodes`
 
-        if [ -n "$deployceph" ] ; then
-            echo "Setting second last node to SLE12 Storage..."
-            set_node_role_and_platform ${nodes[0]} "storage" "suse-12.0"
+        for systems in ${want_node_os//,/ } ; do
+            node_os=${systems%=*}
+            number=${systems#*=}
+            i=1
```
Comment (aplanas) [inline scripts/qa_crowbarsetup.sh]: `local`s ?

File: scripts/qa_crowbarsetup.sh [hunk 6]
```diff
@@ -1648,35 +1674,67 @@ function onadmin_allocate()
...
-            $(get_all_discovered_nodes | tail -n 2)
-        )
+        local nodesavailable=`get_all_discovered_nodes`
 
-        if [ -n "$deployceph" ] ; then
-            echo "Setting second last node to SLE12 Storage..."
-            set_node_role_and_platform ${nodes[0]} "storage" "suse-12.0"
+        for systems in ${want_node_os//,/ } ; do
+            node_os=${systems%=*}
+            number=${systems#*=}
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: Just pushing @aplanas's comment further: no `local` here?

File: scripts/qa_crowbarsetup.sh [hunk 7]
```diff
@@ -1648,35 +1674,67 @@ function onadmin_allocate()
...
-    if [ -n "$wanthyperv" ] ; then
-        echo "Setting last node to Hyper-V compute..."
-        local computenode=$(get_all_discovered_nodes | tail -n 1)
-        set_node_role_and_platform $computenode "compute" "hyperv-6.3"
+        local nodesavailable=`get_all_discovered_nodes`
+
+        for roles in ${want_node_roles//,/ } ; do
+            role=${roles%=*}
+            number=${roles#*=}
+            i=1
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: Same here.

**Comment by vuntz** [comment]
+1

**Pull request merged**